### PR TITLE
Add C++ Api For Cronet

### PR DESCRIPTION
**Pull request by gcasto**

### CODE REVIEW

**Comment by gcasto** [comment]
This is a first pass at adding a C++ API for Cronet. I still haven't figured out how testing should work.

**Comment by makdharma** [comment]
Hi - the makefile is not updated directly but it is generated. Adding @nicolasnoble to advise on how to get this functionality checked in, and adding @yang-g for reviewing the API.

**Comment by gcasto** [comment]
I did generate the Makefile using "./tools/buildgen/generate_projects.sh". It's possible that I used it script incorrectly though. Let me see see what happened.

**Comment by murgatroid99** [comment]
this is ok to test

It looks to me like the Makefile changes were properly generated from the build.yaml changes. We'll have to see the test output to be sure, but I think that part is OK.

**Comment by gcasto** [comment]
BTW, I tried running the tests locally as specified by the [contributing documentation](https://github.com/grpc/grpc/blob/master/CONTRIBUTING.md), but it didn't seem to work.

gcasto@gcasto-2:~/grpc$ ./tools/run_tests/run_tests.py 
languages with custom make options cannot be built simultaneously with other languages

I could run tests for a specific language using the "-l" flag, but I couldn't run the entire suite. Is this expected?

**Comment by murgatroid99** [comment]
It looks like this change would only affect the C++ target, so you should only need to run the C++ tests to verify it.

**Comment by gcasto** [comment]
So, it looks like the tests are failing, but I don't think that this should be related to my change. This shouldn't even change the compilation on Window for example. Should I do something to re-run the tests?

File: src/cpp/client/cronet_credentials.cc [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      const string& target, const grpc::ChannelArguments& args) GRPC_OVERRIDE {
+    grpc_channel_args channel_args;
+    args.SetChannelArgs(&channel_args);
+    return CreateChannelInternal(
+        "",
+        grpc_cronet_secure_channel_create(
+            engine_, target.c_str(), &channel_args, nullptr));
+  }
+
+  // TODO(gcasto): Should this derive from SecureChannelCredentials?
```
Comment (yang-g) [inline src/cpp/client/cronet_credentials.cc]: From what I can see, this cronet channel looks like unsecure. This also means it does not work with other existing credentials types.

**Comment by yang-g** [comment]
LGTM

**Comment by gcasto** [comment]
What's the next step here? It looks like I still need approval.

**Comment by yang-g** [comment]
jenkins: test this please

**Comment by gcasto** [comment]
Are these flakes in the tests? These don't seem related to my change (it shouldn't even have touched the Windows).

**Comment by makdharma** [comment]
@nicolasnoble @jtattermusch - Could you please look at the failures? They indeed seem unrelated.

**Comment by jtattermusch** [comment]
@makdharma   no point for me too look at the tests when one of the most important suites "Basic Tests Linux" didn't finish running (it probably got stuck for issues associated with worker reboots, so you'll need to restart the tests).

**Comment by jtattermusch** [comment]
Please rebase to get a fresh test run and then we can look at the results.

**Comment by gcasto** [comment]
Looks like the tests are still failing.

**Comment by gcasto** [comment]
Ping?

**Comment by yang-g** [comment]
Could you add copyright header in: src/cpp/client/cronet_credentials.cc
and clang-format the code according to:
--- /local-code/src/cpp/client/cronet_credentials.cc	2016-10-17 17:05:47.590557067 +0000
+++ /tmp/tmp.b7466Z2NzX	2016-10-17 17:40:06.186644296 +0000
@@ -1,32 +1,30 @@
 #include <grpc++/security/credentials.h>
 
-#include <grpc/grpc_cronet.h>
 #include <grpc++/channel.h>
 #include <grpc++/support/channel_arguments.h>
+#include <grpc/grpc_cronet.h>
 #include "src/cpp/client/create_channel_internal.h"
 
 namespace grpc {
 
 class CronetChannelCredentialsImpl GRPC_FINAL : public ChannelCredentials {
  public:
-  CronetChannelCredentialsImpl(void* engine)
-      : engine_(engine) {}
+  CronetChannelCredentialsImpl(void* engine) : engine_(engine) {}
 
   std::shared_ptr<grpc::Channel> CreateChannel(
       const string& target, const grpc::ChannelArguments& args) GRPC_OVERRIDE {
     grpc_channel_args channel_args;
     args.SetChannelArgs(&channel_args);
     return CreateChannelInternal(
-        "",
-        grpc_cronet_secure_channel_create(
-            engine_, target.c_str(), &channel_args, nullptr));
+        "", grpc_cronet_secure_channel_create(engine_, target.c_str(),
+                                              &channel_args, nullptr));
   }
 
   SecureChannelCredentials* AsSecureCredentials() GRPC_OVERRIDE {
     return nullptr;
   }
 
-private:
+ private:
   void* engine_;
 };

**Comment by yang-g** [comment]
Better formatted:
```
--- /local-code/src/cpp/client/cronet_credentials.cc	2016-10-17 17:05:47.590557067 +0000
+++ /tmp/tmp.b7466Z2NzX	2016-10-17 17:40:06.186644296 +0000
@@ -1,32 +1,30 @@
 #include <grpc++/security/credentials.h>
 
-#include <grpc/grpc_cronet.h>
 #include <grpc++/channel.h>
 #include <grpc++/support/channel_arguments.h>
+#include <grpc/grpc_cronet.h>
 #include "src/cpp/client/create_channel_internal.h"
 
 namespace grpc {
 
 class CronetChannelCredentialsImpl GRPC_FINAL : public ChannelCredentials {
  public:
-  CronetChannelCredentialsImpl(void* engine)
-      : engine_(engine) {}
+  CronetChannelCredentialsImpl(void* engine) : engine_(engine) {}
 
   std::shared_ptr<grpc::Channel> CreateChannel(
       const string& target, const grpc::ChannelArguments& args) GRPC_OVERRIDE {
     grpc_channel_args channel_args;
     args.SetChannelArgs(&channel_args);
     return CreateChannelInternal(
-        "",
-        grpc_cronet_secure_channel_create(
-            engine_, target.c_str(), &channel_args, nullptr));
+        "", grpc_cronet_secure_channel_create(engine_, target.c_str(),
+                                              &channel_args, nullptr));
   }
 
   SecureChannelCredentials* AsSecureCredentials() GRPC_OVERRIDE {
     return nullptr;
   }
 
-private:
+ private:
   void* engine_;
 };

```

**Comment by gcasto** [comment]
Still not sure if I understand what is going on with the tests. Any suggestions?

**Comment by yang-g** [comment]
I think they are flaky tests. Merging.

**Pull request merged**